[api] Rest API for PKI app

ManishShah120 opened the pull request:
Closes #462

atb00ker commented on openwisp_controller/pki/api/views.py:
```diff
@@ -0,0 +1,76 @@
+    queryset = Ca.objects.all()
+
+
+def crl_download_view(request, pk):
```
Isn't this function doing the same job as `[django-x509 crl_view](https://github.com/openwisp/django-x509/blob/master/django_x509/base/admin.py#L135-L143)`?
Can we please try to avoid duplicating this code? :smile:

ManishShah120 replied:
Yes, I tried to import it, but didn't work. I will try again. :+1:

atb00ker replied:
Remember you can change the code in django-x509 to move the required function out of that class and create a mixin! :smile:

ManishShah120 replied:
Hi, @atb00ker I tried creating a mixin in django-x509, and it works well, but when I import it in drf view I get error's regarding arguments.

atb00ker replied:
Sorry, I missed this message in mail.
Sure, I can help you out with it. Can you share the stacktrace that you are seeing? `error's regarding arguments` is not enough to guess what the error is ...

atb00ker replied:
Meanwhile, I wrote this working code:
```
from django_x509 import settings as app_settings
from django.http import HttpResponse
from django.shortcuts import get_object_or_404
from django.utils.translation import ugettext_lazy as _
from django.views import View
from ..models import Ca


class CrlDownloadMixin:
    def crl_view(self, request, pk):
        authenticated = request.user.is_authenticated
        authenticated = authenticated() if callable(authenticated) else authenticated
        if app_settings.CRL_PROTECTED and not authenticated:
            return HttpResponse(_('Forbidden'), status=403, content_type='text/plain')
        instance = get_object_or_404(self.model, pk=pk)
        return HttpResponse(
            instance.crl, status=200, content_type='application/x-pem-file'
        )


class CrlDownloadMixinView(View, CrlDownloadMixin):
    model = Ca

    def get(self, request, pk):
        return self.crl_view(request, pk)


crlview = CrlDownloadMixinView.as_view()
```

```
class AbstractCaAdmin(BaseAdmin, CrlDownloadMixin):
   ....
```

```
from .base.views import crlview
from django.urls import path

urlpatterns = [path('api/<pk:int>', crlview, name='crl')]
```

Note: this is only a rough implementation using model as is and using random url path, please polish it before using.

ManishShah120 replied:
Thanx a Lot. Done implementing

atb00ker reviewed (commented):
Going in the correct direction, one question:

nemesisdesign commented on openwisp_controller/pki/api/views.py:
```diff
@@ -45,11 +49,15 @@ class CaDetailView(ProtectedAPIMixin, RetrieveUpdateDestroyAPIView):
+        instance = get_object_or_404(Ca, pk=kwargs['pk'])
+        return HttpResponse(
+            instance.crl, status=200, content_type='application/x-pem-file'
+        )
```
@ManishShah120 have you made sure the CRL view from django-x509 is not used?

ManishShah120 replied:
Yes, I checked it. It's not being used.

nemesisdesign commented on openwisp_controller/pki/api/serializers.py:
```diff
@@ -0,0 +1,184 @@
+        initial='sha256',
+        label='Digest algorithm',
+        required=True,
+    )
```
why do we have to redefine this and the other choice fields? Is it really necessary?

ManishShah120 replied:
Hi I redefined it to exclude the unwanted choice field i.e., 'blank' and 'none'. But this will get resolved with this. :- https://github.com/openwisp/openwisp-controller/issues/444 So simplefied the code.

nemesisdesign commented on openwisp_controller/pki/tests/test_api.py:
```diff
@@ -0,0 +1,153 @@
+        with self.assertNumQueries(8):
+            response = self.client.delete(path)
+        self.assertEqual(response.status_code, 204)
+        self.assertEqual(Cert.objects.count(), 0)
```
please add `assertNumQueries` to each test to ensure we're keeping queries in check

ManishShah120 replied:
Done adding. :+1:

atb00ker reviewed (commented):
Looks good, need to do functional testing, will pick it up on the weekend, meanwhile, please continue with connections. :smile:

atb00ker commented on openwisp_controller/pki/api/serializers.py:
```diff
@@ -0,0 +1,144 @@
+            'notes',
+            'key_length',
+            'digest',
+            'validity_start',
+            'validity_end',
```
The default values are giving this error: 
```

{
    "validity_start": [
        "Datetime has wrong format. Use one of these formats instead: YYYY-MM-DDThh:mm[:ss[.uuuuuu]][+HH:MM|-HH:MM|Z]."
    ],
    "validity_end": [
        "Datetime has wrong format. Use one of these formats instead: YYYY-MM-DDThh:mm[:ss[.uuuuuu]][+HH:MM|-HH:MM|Z]."
    ]
}
```
Please fix the defaults.

ManishShah120 replied:
@atb00ker The second error is related with model, admin I raised an issue here:- https://github.com/openwisp/openwisp-controller/issues/444 this will get resolved with that.

atb00ker reviewed (changes requested):
Hi, 

Are you sure you have tested all the edge cases, I tried only one url, `/api/v1/pki/ca/` just gave it title and it failed, fixed that error, another failure.

Please double check all the urls are working for edge cases before submitting for review.

second error:
![2021-06-19_21-11](https://user-images.githubusercontent.com/24985760/122647793-92bff600-d143-11eb-8c35-c2ebc6c419ea.png)

ManishShah120 commented:
Added the endpoint for importing existing Ca

ManishShah120 commented:
I just pushed the code, Will try to fix the changes requested on gitter ASAP.

ManishShah120 commented:
Added the import of existing Ca/Cert to list API endpoints.

ManishShah120 commented:
Hi, I have tested this PR and found one minor issues with the date time format in the djangox509 module itself, the format of validity end, is not same as that of validity start, However this is not causing any issue.

atb00ker commented:
Hi Manish, I checked the PKI:
1. https://github.com/openwisp/openwisp-controller/pull/455#discussion_r654810018 is still failing, do I need to update some module to see the changes?
2. https://github.com/openwisp/openwisp-controller/pull/455#issuecomment-871327293
What issue are you talking about? What changes do we need to make in django-x509? I don't see a PR there, is it in progress?

Moved this back to in-progress?

ManishShah120 commented:
> Hi Manish, I checked this:
> 
> 1. [[api] Rest API for PKI app #455 (comment)](https://github.com/openwisp/openwisp-controller/pull/455#discussion_r654810018) is still failing, ~do I need to update some module to see the changes~? Checked it with all latest, how is it working for you? I have to change the Datetime to make it work.

Hi, I am not able to replicate this, and it's working for fine for me, here is  gif:-
![workingrestapipki](https://user-images.githubusercontent.com/45564404/124080770-5e640800-da68-11eb-9914-6b7dd4a91c9b.gif)




> 2. [[api] Rest API for PKI app #455 (comment)](https://github.com/openwisp/openwisp-controller/pull/455#issuecomment-871327293)
>    What issue are you talking about? [[pki] CA can't be created when `Key length` set to `None` or `Digest algorithm` set to `None` #444](https://github.com/openwisp/openwisp-controller/issues/444)? Have you started with it?

No, Haven't started it, this options I have kept same as the admin panel. :smile:

atb00ker commented:
Okay, it worked on chrome and failed on firefox. This is odd.
Perhaps we can continue for now but create a low priority issue and check out the reason for it! :smile:

atb00ker commented on openwisp_controller/pki/api/urls.py:
```diff
@@ -0,0 +1,25 @@
+            path('ca/', api_views.ca_list, name='ca_list'),
+            path('ca/<str:pk>/', api_views.ca_detail, name='ca_detail'),
+            path('ca/<str:pk>/crl', api_views.crl_download, name='crl_download'),
+            path('cert/', api_views.cert_list, name='cert_list'),
```
| SN | /api/v1/pki/cert/                             | Status |                                Details                                |
|:--:|-----------------------------------------------|:------:|:---------------------------------------------------------------------:|
|  1 | [SU][O1A][O1O] Can view only appropriate ca   |  Pass  |                                                                       |
|  2 | [SU][O1A][O1O] Can create only appropriate ca |  Fail  | When trying to import certificate, I need to provide validity period. |
|  3 | [SU][O1A] All errors reported together        |  Pass  |                                                                       |

ManishShah120 replied:
> When trying to import certificate, I need to provide validity period.

@atb00ker By this do you mean `validity_start` & `validity_end`, this fields are not required fields, and even if the data get''s sent in a POST request, it gets replaced with the `cert` data from which it is imported.

atb00ker replied:
@ManishShah120 Are you sure they are not required? It wouldn't let me submit on firefox without these fields!

ManishShah120 replied:
Yes, I am able to submit it, I also tested it in JSON format form.

nemesisdesign replied:
I confirm that I have been able to import as expected and without issues on Chrome,
while on Firefox as @atb00ker noted, the validity start and end fields are somehow prefilled and hence submitting the form fails with a validation error unless the prefilled value is edited manually.

Not a big deal, because this will happen only when using it from the browsable API. @ManishShah120 can you double check on Firefox and see if you can force the default value to be empty there too?

atb00ker commented on openwisp_controller/pki/api/urls.py:
```diff
@@ -0,0 +1,25 @@
+            path('ca/<str:pk>/', api_views.ca_detail, name='ca_detail'),
+            path('ca/<str:pk>/crl', api_views.crl_download, name='crl_download'),
+            path('cert/', api_views.cert_list, name='cert_list'),
+            path('cert/<str:pk>/', api_views.cert_detail, name='cert_detail'),
```
| SN | /api/v1/pki/cert/str:pk                       | Status | Details |
|----|-----------------------------------------------|:------:|:-------:|
| 1  | [SU][O1A][O1O] Can view only appropriate Cert |  Pass  |         |
| 2  | [SU][O1A][O1O] Can edit only appropriate Cert |  Pass  |         |
| 3  | [SU][O1A] Correct data is shown               |  Pass  |         |
| 4  | [SU][01A] Can delete only appropriate data    |  Pass  |         |

atb00ker reviewed (commented):
I see one critical failure.

nemesisdesign commented on README.rst:
```diff
@@ -898,6 +898,114 @@ Delete VPN
+
+.. code-block:: text
+
+    GET /api/v1/pki/ca/
```
as shared in the gitter chat, let's follow the URL naming consistently with the other sub apps, so `/api/v1/<module-name>/<resource-name>/` which translates to `/api/v1/controller/ca/`, applies to other occurrences.

nemesisdesign commented on README.rst:
```diff
@@ -898,6 +898,114 @@ Delete VPN
+
+.. code-block:: text
+
+    GET /api/v1/pki/ca/{id}/crl
```
final slash missing

nemesisdesign commented on README.rst:
```diff
@@ -898,6 +898,114 @@ Delete VPN
+
+    GET /api/v1/pki/cert/
+
+Create Cert
```
I'd specify "Create new Cert"

nemesisdesign commented on README.rst:
```diff
@@ -898,6 +898,114 @@ Delete VPN
+
+Import existing Cert
+^^^^^^^^^^^^^^^^^^^^
+
```
can you please spend a few words here to explain how to import a cert (specify the fields required to make the import cert work)?

nemesisdesign commented on README.rst:
```diff
@@ -898,6 +898,114 @@ Delete VPN
+
+    GET /api/v1/pki/ca/
+
+Create CA
```
I'd specify "Create new CA"

nemesisdesign commented on README.rst:
```diff
@@ -898,6 +898,114 @@ Delete VPN
+
+Import existing CA
+^^^^^^^^^^^^^^^^^^
+
```
can you please spend a few words here to explain how to import a CA (specify the fields required to make the import cert work)?

nemesisdesign reviewed (changes requested):
Great progress @ManishShah120 @atb00ker  :+1: :clap:.

See my comments below.

ManishShah120 commented:
Hi @nemesisdesign @atb00ker, all the requested changes have been resolved, please have a look. :blush:

nemesisdesign reviewed (approved):
Tested again with Firefox and I confirm that problem @atb00ker reported is fixed.

LGTM!

okraits reviewed (commented):
Tests went fine, no findings except my little comments.

ManishShah120 commented:
@okraits Fixed the requested changes. :+1:

ManishShah120 commented on README.rst:
```diff
@@ -898,6 +900,122 @@ Delete VPN
+    POST /api/v1/controller/cert/
+
+**Note**: To import an existing Cert, only ``name``, ``ca``,
+``certificate`` and ``private_key`` fields has to be filled
```
Done.

ManishShah120 commented on README.rst:
```diff
@@ -898,6 +900,122 @@ Delete VPN
+    POST /api/v1/controller/ca/
+
+**Note**: To import an existing CA, only ``name``, ``certificate``
+and ``private_key`` fields has to be filled in the ``HTML`` form or
```
Done

okraits commented:
> @okraits Fixed the requested changes. +1

Thank you! :grinning:

atb00ker reviewed (dismissed):
Found a very minor problem, when I create a new cert and press save, the page is reloaded showing the data of the newly created cert and if I try to create a new cert, the `Extensions` value is pre-filled as `null`, I need to change it to a valid json: `{}` to continue working.
Although I doubt this will be a problem for a real user.

nemesisdesign commented:
> Found a very minor problem, when I create a new cert and press save, the page is reloaded showing the data of the newly created cert and if I try to create a new cert, the `Extensions` value is pre-filled as `null`, I need to change it to a valid json: `{}` to continue working.
> Although I doubt this will be a problem for a real user.

ok would be good to double check this @ManishShah120, let us know if it's something that can be fixed.

ManishShah120 commented:
> Found a very minor problem, when I create a new cert and press save, the page is reloaded showing the data of the newly created cert and if I try to create a new cert, the `Extensions` value is pre-filled as `null`, I need to change it to a valid json: `{}` to continue working.
> Although I doubt this will be a problem for a real user.

Hi, @atb00ker this can be fixed, this happens because thie field I have kept it as `write_only` as done or is in the admin panel, to fix this i.e., give this field both write and read permissions, but since after creating a new `Cert` in the admin panel, the `extension` field is not displayed so I thought it to keep it as `write_only` field.

Should I give it both `read` and `write` permisison, but in the admin panel it is set as `write_only` field. :smile:

nemesisdesign commented:
> > Found a very minor problem, when I create a new cert and press save, the page is reloaded showing the data of the newly created cert and if I try to create a new cert, the `Extensions` value is pre-filled as `null`, I need to change it to a valid json: `{}` to continue working.
> > Although I doubt this will be a problem for a real user.
> 
> Hi, @atb00ker this can be fixed, this happens because thie field I have kept it as `write_only` as done or is in the admin panel, to fix this i.e., give this field both write and read permissions, but since after creating a new `Cert` in the admin panel, the `extension` field is not displayed so I thought it to keep it as `write_only` field.
> 
> Should I give it both `read` and `write` permisison, but in the admin panel it is set as `write_only` field.

This field can be read.

nemesisdesign reviewed (changes requested):
@ManishShah120 why in the cert detail endpoint the name of the CA is shown while in the list endpoint it's the ID?

I would always show the ID because it's the only way to be super sure someone is dealing with a specific object.

Let's be consistent.

ManishShah120 commented:
Fixed:-
- Ca field ID (not showing in the cert detail endpoint)
- extensions fields returned as `null` after post requests.
- added tests for the above changes.

ManishShah120 commented:
> @ManishShah120 why in the cert detail endpoint the name of the CA is shown while in the list endpoint it's the ID?
> 
> I would always show the ID because it's the only way to be super sure someone is dealing with a specific object.
> 
> Let's be consistent.

@nemesisdesign, I tried to keep it similar to the admin panel, But, Fixed it now. :blush:

nemesisdesign reviewed (changes requested):
@ManishShah120 I merged the latest master in and some query checks failed can you please double check?

ManishShah120 commented:
> @ManishShah120 I merged the latest master in and some query checks failed can you please double check?

Fixed it :+1: @nemesisdesign

atb00ker reviewed (approved):
LGTM! :smile:

nemesisdesign reviewed (changes requested):
So here's some important actions that can be performed from the admin but seem missing from the API.

- renew cert
- renew CA
- revoke certificate

To keep it simple, we can introduce these actions for singular objects (leaving the implementation for performing renew/revoke on multiple objects for a future phase if needed at all).

The models have method for each of these actions so the implementation should be pretty easy.

The API urls can be like:

POST /cert/{id}/renew/
POST /ca/{id}/renew/
POST /cert/{id}/revoke/

nemesisdesign commented on openwisp_controller/pki/api/views.py:
```diff
@@ -49,6 +52,21 @@ class CaDetailView(ProtectedAPIMixin, RetrieveUpdateDestroyAPIView):
+        instance.renew()
+        return Response(
+            {_("CA '{}' renewed successfully".format(instance.name))}, status=302,
+        )
```
return the serialized data with the updated data and 200 as status code, applies to the other endpoints as well

ManishShah120 replied:
This is done. :+1:

nemesisdesign reviewed (changes requested):
almost there

ManishShah120 commented:
@nemesisdesign I have made the requested changes for the revoke and renew endpoints, but there's one minor issue, i.e., I have used the `CaDetailSerializer` & `CertDetailSerializer` for serializing the instance data for returning it, but since in this serializers some of the fields are set to `write` so these fields are also displayed in the response page. 

However, this will not be a an issue, in production. Still I will see if I can fix it.

nemesisdesign commented:
> @nemesisdesign I have made the requested changes for the revoke and renew endpoints, but there's one minor issue, i.e., I have used the `CaDetailSerializer` & `CertDetailSerializer` for serializing the instance data for returning it, but since in this serializers some of the fields are set to `write` so these fields are also displayed in the response page.
> 
> However, this will not be a an issue, in production. Still I will see if I can fix it.

Sure, better fix this and ensure the serializers are correct, you simply need to create a new serializers inheriting the serializer you're using now and removing the write_only fields.

ManishShah120 commented:
> Sure, better fix this and ensure the serializers are correct, you simply need to create a new serializers inheriting the serializer you're using now and removing the write_only fields.

Done, Fixed it. :+1:

nemesisdesign commented on openwisp_controller/pki/api/serializers.py:
```diff
@@ -0,0 +1,196 @@
+    return data
+
+
+def CertList_fields(fields=None):
```
two doubts here:

- why the default argument is `None`? it looks like there should be no default argument
- why this naming? Capital letter is used for classes, I think this should be named `get_cert_list_fields`

Eg:

```python
def get_cert_list_fields(fields):
    pass
```

nemesisdesign commented on openwisp_controller/pki/api/serializers.py:
```diff
@@ -0,0 +1,196 @@
+        return value
+
+
+def CaDetail_fields(fields=None):
```
same here as for `get_cert_list_fields`

nemesisdesign commented on openwisp_controller/pki/api/serializers.py:
```diff
@@ -0,0 +1,196 @@
+        return value
+
+
+def CertDetail_fields(fields=None):
```
same here

nemesisdesign commented on openwisp_controller/pki/api/serializers.py:
```diff
@@ -0,0 +1,186 @@
+    def validate_validity_end(self, value):
+        if value is None:
+            value = default_cert_validity_end()
+        return value
```
can you please move these 3 methods above to a `BaseListSerializer` which inherits from `BaseSerializer` and then use it as a base for `CertListSerializer` and `CatListSerializer`?

The line which adds `data.update({'ca': instance.ca})` can be added with super()

nemesisdesign reviewed (changes requested):
Almost there, let's do some more code improvement and then we're done.

ManishShah120 commented:
Hi @nemesisdesign have implemented the `BaseListSerializer`, Is the implementation correct.

nemesisdesign reviewed (changes requested):
I found another weird thing: why does the list endpoint provide the info about extensions but the details endpoint does not?

ManishShah120 commented:
> I found another weird thing: why does the list endpoint provide the info about extensions but the details endpoint does not?

@nemesisdesign currently I have also removed the `passphrase` field from the detail endpoint, Should I leave it as as it is or should I make it visible too in the detail endpoint.

nemesisdesign reviewed (approved):
> > I found another weird thing: why does the list endpoint provide the info about extensions but the details endpoint does not?
> 
> @nemesisdesign currently I have also removed the `passphrase` field from the detail endpoint, Should I leave it as as it is or should I make it visible too in the detail endpoint.

Passphrase shall be `write_only`: https://github.com/openwisp/openwisp-controller/pull/455/commits/a020222684a67f151c6f0730791eea96eceef81f.

With this change I think we're done here.
